### PR TITLE
Stat

### PR DESCRIPTION
**Pull request by bretambrose**
WIP

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/io/channel_bootstrap.h [hunk 1]
```diff
@@ -25,31 +25,22 @@ struct aws_socket_options;
...
+ *   (2) Channel setup - If TLS is being used, this function is called once the socket has connected, the channel has
+ * been initialized, and TLS has been successfully negotiated. A TLS handler has already been added to the channel. If
+ * TLS negotiation fails, this function will be called with the corresponding error code. If TLS is not being used, this
+ * function is called once the socket has connected and the channel has been initialized.
+ *   (3) Channel shutdown
  *
- * This function is always called within the thread of the event-loop that the new channel is assigned to.
+ * These callbacks are always invoked within the thread of the event-loop that the channel is assigned to.
  *
  * This function does NOT mean "success", if error_code is AWS_OP_SUCCESS then everything was successful, otherwise an
```
Comment (justinboswell) [inline include/aws/io/channel_bootstrap.h]: s/mean/always imply/

File: include/aws/io/statistics.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    AWSCRT_STAT_CAT_TLS
+};
+
+struct aws_crt_statistics_socket {
+    aws_crt_statistics_category_t category;
+    uint64_t bytes_read;
+    uint64_t bytes_written;
+};
+
+enum aws_tls_negotiation_status {
```
Comment (justinboswell) [inline include/aws/io/statistics.h]: This feels like it should be in the tls_handler.h, or does this just never get used by anything but stats?

File: source/channel.c [hunk 3]
```diff
@@ -963,3 +997,103 @@ size_t aws_channel_handler_initial_window_size(struct aws_channel_handler *handl
...
+    uint64_t now_ms = aws_timestamp_convert(now_ns, AWS_TIMESTAMP_NANOS, AWS_TIMESTAMP_MILLIS, NULL);
+
+    struct aws_array_list *statistics_list = &channel->statistic_list;
+    aws_array_list_clear(statistics_list);
+
+    struct aws_channel_slot *current_slot = channel->first;
+    while (current_slot) {
+        struct aws_channel_handler *handler = current_slot->handler;
+        if (handler != NULL && handler->vtable->append_statistics != NULL) {
+            handler->vtable->append_statistics(handler, statistics_list);
```
Comment (justinboswell) [inline source/channel.c]: Should this be ```gather_statistics()``` for consistency?

File: source/darwin/secure_transport_tls_channel_handler.c [hunk 4]
```diff
@@ -321,6 +325,10 @@ static void s_invoke_negotiation_callback(struct aws_channel_handler *handler, i
 static int s_drive_negotiation(struct aws_channel_handler *handler) {
     struct secure_transport_handler *secure_transport_handler = handler->impl;
 
+    if (secure_transport_handler->stats.handshake_status == AWS_MTLS_STATUS_NONE) {
```
Comment (justinboswell) [inline source/darwin/secure_transport_tls_channel_handler.c]: Should we assert if it's in another state? Is that even possible?

File: source/s2n/s2n_tls_channel_handler.c [hunk 5]
```diff
@@ -325,14 +327,34 @@ static int s_s2n_handler_send(void *io_context, const uint8_t *buf, uint32_t len
...
+    }
+
+    s2n_handler->stats.handshake_status =
+        (error_code == AWS_ERROR_SUCCESS) ? AWS_MTLS_STATUS_SUCCESS : AWS_MTLS_STATUS_FAILURE;
+}
+
 static int s_drive_negotiation(struct aws_channel_handler *handler) {
     struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
 
+    if (s2n_handler->stats.handshake_status == AWS_MTLS_STATUS_NONE) {
```
Comment (justinboswell) [inline source/s2n/s2n_tls_channel_handler.c]: Same here, should we assert on this? Could it be ```SUCCESS```/```ERROR```, and then we're trying again?

File: source/windows/secure_channel_tls_handler.c [hunk 6]
```diff
@@ -1551,15 +1581,15 @@ struct aws_byte_buf aws_tls_handler_server_name(struct aws_channel_handler *hand
...
-static struct aws_channel_handler_vtable s_handler_vtable = {
-    .destroy = s_handler_destroy,
-    .process_read_message = s_process_read_message,
-    .process_write_message = s_process_write_message,
-    .shutdown = s_handler_shutdown,
-    .increment_read_window = s_increment_read_window,
-    .initial_window_size = s_initial_window_size,
-    .message_overhead = s_message_overhead,
-};
+static struct aws_channel_handler_vtable s_handler_vtable = {.destroy = s_handler_destroy,
```
Comment (justinboswell) [inline source/windows/secure_channel_tls_handler.c]: Did clang-format allow this?!

File: include/aws/io/statistics.h [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    AWS_MTLS_STATUS_ONGOING,
+    AWS_MTLS_STATUS_SUCCESS,
+    AWS_MTLS_STATUS_FAILURE
+};
+
+/**
+ * Tls channel handler statistics record
+ */
+struct aws_crt_statistics_tls {
+    aws_crt_statistics_category_t category;
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply include/aws/io/statistics.h]: it's a uint32_t typedef, there as a semantic reminder.  The actual values come from a per-library enum.  Similar to error code and log subject.

File: include/aws/io/statistics.h [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+};
+
+/**
+ * An enum for the current state of tls negotiation within a tls channel handler
+ */
+enum aws_tls_negotiation_status {
+    AWS_MTLS_STATUS_NONE,
+    AWS_MTLS_STATUS_ONGOING,
+    AWS_MTLS_STATUS_SUCCESS,
+    AWS_MTLS_STATUS_FAILURE
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply include/aws/io/statistics.h]: Definitely

File: source/windows/secure_channel_tls_handler.c (hunk 6 above)
Comment (bretambrose) [reply source/windows/secure_channel_tls_handler.c]: Missing comma =/

File: include/aws/io/statistics.h (hunk 2 above)
Comment (bretambrose) [reply include/aws/io/statistics.h]: Moved

File: source/darwin/secure_transport_tls_channel_handler.c (hunk 4 above)
Comment (bretambrose) [reply source/darwin/secure_transport_tls_channel_handler.c]: I guess we could verify it's always in either None or Ongoing.

File: source/s2n/s2n_tls_channel_handler.c (hunk 5 above)
Comment (bretambrose) [reply source/s2n/s2n_tls_channel_handler.c]: Reworked and added asserts

File: source/statistics.c [hunk 9]
```diff
@@ -44,5 +47,8 @@ void aws_crt_statistics_tls_cleanup(struct aws_crt_statistics_tls *stats) {
 }
 
 void aws_crt_statistics_tls_reset(struct aws_crt_statistics_tls *stats) {
+    /*
+     * We currently don't have any resetable tls statistics yet, but they may be added in the future.
```
Comment (justinboswell) [inline source/statistics.c]: [nitpick] Reset-able

File: include/aws/io/socket.h [hunk 10]
```diff
@@ -114,6 +114,7 @@ struct aws_socket {
     struct aws_socket_options options;
     struct aws_io_handle io_handle;
     struct aws_event_loop *event_loop;
+    struct aws_channel_handler *handler;
```
Comment (JonathanHenson) [inline include/aws/io/socket.h]: why!? aws_socket shouldn't need to know about the handler. Is this leftover from a previous iteration?

File: include/aws/io/statistics.h [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+ */
+
+#include <aws/io/io.h>
+
+#include <aws/common/statistics.h>
+#include <aws/io/tls_channel_handler.h>
+
+enum aws_crt_io_statistics_category {
+    AWSCRT_STAT_CAT_SOCKET = AWS_CRT_STATISTICS_CATEGORY_BEGIN_RANGE(AWS_C_IO_PACKAGE_ID),
+    AWSCRT_STAT_CAT_TLS
```
Comment (JonathanHenson) [inline include/aws/io/statistics.h]: comma

File: include/aws/io/socket.h (hunk 10 above)
Comment (bretambrose) [reply include/aws/io/socket.h]: There's no way to go from s_on_socket_write_complete to the handler otherwise

**Pull request merged**